README.md
# mpe
sample size calculations for multiple co-primary endppoints


mpe/NEWS
###############################################################################
## NEWS to package mpe
###############################################################################

###########################################################
## Version 1.0
###########################################################
- first stable and release version of the package


mpe/R/mpe.t.test.R
mpe.t.test <- function(X, Y, conf.level = 0.975){
  if(missing(X))
    stop("'X' must be given")
  if(missing(Y))
    stop("'Y' must be given")
  if(ncol(X) !=  ncol(Y))
    stop("'X' and 'Y' must have the same number of columns")
  if (!missing(conf.level) && (length(conf.level) != 1 || !is.finite(conf.level) ||
                               conf.level < 0 || conf.level > 1))
    stop("'conf.level' must be a single number between 0 and 1")
  dname <- paste(deparse(substitute(x)), "and", deparse(substitute(y)))
  ncols <- ncol(X)
  pvals <- numeric(ncols)
  statistic <- numeric(ncols)
  parameter <- numeric(ncols)
  conf <- matrix(NA, ncol = 2, nrow = ncols)
  estimate <- matrix(NA, ncol = 2, nrow = ncols)
  for(i in seq_len(ncols)){
    temp <- t.test(X[,i], Y[,i], var.equal = TRUE,
                   conf.level = conf.level)
    pvals[i] <- temp$p.value
    statistic[i] <- temp$statistic
    parameter[i] <- temp$parameter
    conf[i,] <- temp$conf.int
    estimate[i,] <- temp$estimate
  }
  colnames(estimate) <- c("X", "Y")
  rownames(estimate) <- paste("EDP", 1:ncols, sep = ".")
  colnames(conf) <- c(paste(1-conf.level), 1)
  rownames(conf) <- paste("EDP", 1:ncols, sep = ".")
  alternative <- "true difference in means is larger than 0 for all endpoints"
  attr(conf, "conf.level") <- conf.level
  method <- "Intersection-union t-test"
  pval <- max(pvals)
  stat <- min(statistic)
  names(stat) <- "t"
  para <- parameter[1]
  names(para) <- "df"
  rval <- list(method = method, statistic = stat, parameter = para, p.value = pval,
               conf.int = conf, estimate = estimate, alternative= alternative, data.name = dname)
  class(rval) <- "mpe.test"
  return(rval)
  
}






mpe/R/mpe.z.test.R
mpe.z.test <- function(X, Y, Sigma, conf.level = 0.975){
  if(missing(X))
    stop("'X' must be given")
  if(missing(Y))
    stop("'Y' must be given")
  if(ncol(X) !=  ncol(Y))
    stop("'X' and 'Y' must have the same number of columns")
  if(missing(Sigma))
    stop("'Sigma' must be given")
  if(nrow(Sigma) != ncol(Sigma))
    stop("covariance matrix 'Sigma' must be quadratic")
  if(nrow(Sigma) != ncol(X))
    stop("covariance matrix has wrong dimension")
  if(max(abs(Sigma - t(Sigma))) > 1e-10)
    stop("matrix 'Sigma' must be symmetric")
  if(!all(eigen(Sigma)$values > 0))
    stop("matrix 'Sigma' must be positive definite")
  if (!missing(conf.level) && (length(conf.level) != 1 || !is.finite(conf.level) ||
                               conf.level < 0 || conf.level > 1))
    stop("'conf.level' must be a single number between 0 and 1")

  dname <- paste(deparse(substitute(x)), "and", deparse(substitute(y)))
  diff <- colMeans(X) - colMeans(Y)
  names(diff) <- paste("EP", 1:ncol(X), sep = ".")
  n <- nrow(X)
  m <- nrow(Y)
  SD <- sqrt(diag(Sigma))
  Z <- diff/(SD*sqrt(1/n+1/m))
  statistic <- min(Z)
  names(statistic) <- "z"
  pval <- max(pnorm(Z, lower.tail = FALSE))
  alpha <- 1 - conf.level
  cint <- rep(qnorm(1 - alpha), ncol(X))
  cint <- cbind(Z - cint, Inf)*SD*sqrt(1/n+1/m)

  colnames(cint) <- c(paste(1-conf.level), 1)
  rownames(cint) <- paste("EP", 1:ncol(X), sep = ".")

  alternative <- "true difference in means is larger than 0 for all endpoints"
  attr(cint, "conf.level") <- conf.level
  method <- "Intersection-union z-test"
  rval <- list(method = method, statistic = statistic, p.value = pval,
               conf.int = cint, estimate = diff,
               alternative = alternative, data.name = dname)
  class(rval) <- "mpe.test"
  return(rval)
}

print.mpe.test <- function (x, digits = getOption("digits"), prefix = "\t", ...)
{
  cat("\n")
  cat(strwrap(x$method, prefix = prefix), sep = "\n")
  cat("\n")
  cat("data:  ", x$data.name, "\n", sep = "")
  out <- character()
  if (!is.null(x$statistic))
    out <- c(out, paste(names(x$statistic), "=", format(signif(x$statistic,
                                                               max(1L, digits - 2L)))))
  if (!is.null(x$parameter))
    out <- c(out, paste(names(x$parameter), "=", format(signif(x$parameter,
                                                               max(1L, digits - 2L)))))
  if (!is.null(x$p.value)) {
    fp <- format.pval(x$p.value, digits = max(1L, digits -
                                                3L))
    out <- c(out, paste("p-value", if (substr(fp, 1L, 1L) ==
                                       "<") fp else paste("=", fp)))
  }
  cat(strwrap(paste(out, collapse = ", ")), sep = "\n")
  if (!is.null(x$alternative)) {
    cat("alternative hypothesis: ", x$alternative, "\n")
  }
  if (!is.null(x$conf.int)) {
    cat(format(100 * attr(x$conf.int, "conf.level")), " percent confidence intervals:\n", sep = "")
    attr(x$conf.int, "conf.level") <- NULL
    print(x$conf.int)
  }
  if (!is.null(x$estimate)) {
    cat("sample estimates:\n")
    print(x$estimate, digits = digits, ...)
  }
  cat("\n")
  invisible(x)
}



mpe/R/power.unknown.var.R
power.unknown.var <- function(K, n = NULL, delta = NULL, Sigma, SD, rho, sig.level = 0.05,
                              power = NULL,M = 10000, min.n = NULL, max.n = NULL,  tol = .Machine$double.eps^0.25, use.uniroot = TRUE)
  {
  ## check of input

  if(missing(K))
    stop("specify the number of co-primary endpoints")
  if(!is.numeric(K))
    stop("'K' must be a natural number > 1")
  K <- as.integer(K)

  if(is.null(n) & is.null(power))
    stop("either 'n' or 'power' must be specified")
  if(!is.null(n) & !is.null(power))

    stop("either 'n' or 'power' must be NULL")
  if(!is.null(n)){
    if(length(n) > 1){
      warning("length of 'n' is greater than 1, only the first entry is used")
      n <- n[1]
    }
    n <- as.integer(n)
  }
  if(!is.null(power)){
    if(power <= 0 | power >= 1)
      stop("power must be in (0, 1)")
  }
  if(is.null(delta))
    stop("expected effect size 'delta' is missing")
  if(length(delta) < 2)
    stop("length of 'delta' < 2: effect for at least two co-primary endpoints is required")
  if(!all(delta > 0))
    stop("all effect sizes need to be positive")
  if(!missing(Sigma)){
    if(nrow(Sigma) != ncol(Sigma))
      stop("covariance matrix 'Sigma' must be quadratic")
    if(nrow(Sigma) != K)
      stop("covariance matrix must have dimension 'K' x 'K'")
    if(max(abs(Sigma - t(Sigma))) > 1e-10)
      stop("matrix 'Sigma' must be symmetric")
  }
  if(missing(Sigma)){
    if(missing(SD) || missing(rho))
      stop("if 'Sigma' is missing 'SD' and 'rho' must be given.")
    if(length(SD) != K)
      stop("length of 'SD' must be equal to 'K'")
    if(length(rho) != 0.5*K*(K-1))
      stop("length of 'rho' must be equal to '0.5*K*(K-1)'")
    Sigma <- matrix(0, nrow = K, ncol = K)
    iter <- 0
    for(i in 1:(K-1)){
      for(j in (i+1):K){
        iter <- iter + 1
        Sigma[i,j] <- rho[iter]*SD[i]*SD[j]
      }
    }
    Sigma <- Sigma + t(Sigma)
    diag(Sigma) <- SD^2
  }
  if(!all(eigen(Sigma)$values > 0))
    stop("matrix 'Sigma' must be positive definite")
  Sigma.cor <- cov2cor(Sigma)

  if(sig.level <= 0 | sig.level >= 1)
    stop("significance level must be in (0, 1)")

  if(length(M) > 1){
    warning("length of 'M' is greater than 1, only the first entry is used")
    M <- M[1]
  }
  M <- as.integer(M)
  if(is.null(n)){
    if(length(min.n) > 1){
      warning("length of 'min.n' is greater than 1, only the first entry is used")
      min.n <- min.n[1]
    }
    min.n <- as.integer(min.n)
    if(length(max.n) > 1){
      warning("length of 'max.n' is greater than 1, only the first entry is used")
      max.n <- max.n[1]
    }
    max.n <- as.integer(max.n)
    if(min.n < 4)
      stop("'min.n' must be >= 4")
    if(min.n >= max.n)
      stop("'min.n' must be < 'max.n'")
  }
  ## calculations
  if(is.null(power)){
    std.effect <- delta/sqrt(diag(Sigma))
    probs <- numeric(M)
    Ws <- rWishart(M, df = n-2, Sigma = Sigma.cor)
    for(i in 1:M){
      Wi <- diag(Ws[,,i])
      ci <- qt(1-sig.level, df = n-2)*sqrt(Wi/(n-2)) - sqrt(n/2)*std.effect
      probs[i] <- pmvnorm(upper = -ci, sigma = Sigma.cor)
    }
    power <- mean(probs)
  }

  if(is.null(n)){
    std.effect <- delta/sqrt(diag(Sigma))
    ssize.fct <- function(n, std.effect, Sigma.cor, power, M, verbose = FALSE){
      probs <- numeric(M)
      Ws <- rWishart(M, df = n-2, Sigma = Sigma.cor)
      for(i in 1:M){
        Wi <- diag(Ws[,,i])
        ci <- qt(1-sig.level, df = n-2)*sqrt(Wi/(n-2)) - sqrt(n/2)*std.effect
        probs[i] <- pmvnorm(upper = -ci, sigma = Sigma.cor) - power
      }
      if(verbose) cat("Current precision:\t", mean(probs), "\n")
      mean(probs)
    }
    if(use.uniroot){
      n <- uniroot(ssize.fct, c(min.n, max.n), tol = tol, extendInt = "yes",
                   std.effect = std.effect, Sigma.cor = Sigma.cor, power = power, M = M,
                   verbose = TRUE)$root
    }else{
      ns <- min.n:max.n
      res <- sapply(ns, ssize.fct, std.effect = std.effect, Sigma.cor = Sigma.cor,
                    power = power, M = M)
      ns.pos <- ns[res > 0]
      res.pos <- res[res > 0]
      cat("Precision:\t", min(res.pos), "\n")
      n <- ns.pos[which.min(res.pos)]
    }
  }
  NOTE <- "n is number in *each* group"
  METHOD <- "Power calculation for multiple co-primary endpoints (covariance unknown)"
  structure(list(n = n, delta = delta, SD = sqrt(diag(Sigma)),
                 rho = Sigma.cor[lower.tri(Sigma.cor)], Sigma = Sigma, sig.level = sig.level,
                 power = power, note = NOTE, method = METHOD),
            class = "power.mpe.test")
}



mpe/R/atleast.one.endpoint.R
atleast.one.endpoint <- function(K, n = NULL, delta = NULL, Sigma, SD, rho, sig.level = 0.05/K,
                                 power = NULL, tol = .Machine$double.eps^0.25){
  ## check of input
  
  if(missing(K))
    stop("specify the number of co-primary endpoints")
  if(!is.numeric(K))
    stop("'K' must be a natural number > 1")
  K <- as.integer(K)
  if(is.null(n) & is.null(power))
    stop("either 'n' or 'power' must be specified")
  if(!is.null(n) & !is.null(power))
    stop("either 'n' or 'power' must be NULL")
  if(!is.null(n)){
    if(length(n) > 1){
      warning("length of 'n' is greater than 1, only the first entry is used")
      n <- n[1]
    }
    n <- as.integer(n)
  }
  if(!is.null(power)){
    if(power <= 0 | power >= 1)
      stop("power must be in (0, 1)")
  }
  if(is.null(delta))
    stop("expected effect size 'delta' is missing")
  if(length(delta) < 2)
    stop("length of 'delta' < 2: effect for at least two co-primary endpoints is required")
  if(!all(delta > 0))
    stop("all effect sizes need to be positive")
  if(!missing(Sigma)){
    if(nrow(Sigma) != ncol(Sigma))
      stop("covariance matrix 'Sigma' must be quadratic")
    if(nrow(Sigma) != K)
      stop("covariance matrix must have dimension 'K' x 'K'")
    if(max(abs(Sigma - t(Sigma))) > 1e-10)
      stop("matrix 'Sigma' must be symmetric")
  }
  if(missing(Sigma)){
    if(missing(SD) || missing(rho))
      stop("if 'Sigma' is missing 'SD' and 'rho' must be given.")
    if(length(SD) != K)
      stop("length of 'SD' must be equal to 'K'")
    if(length(rho) != 0.5*K*(K-1))
      stop("length of 'rho' must be equal to '0.5*K*(K-1)'")
    Sigma <- matrix(0, nrow = K, ncol = K)
    iter <- 0
    for(i in 1:(K-1)){
      for(j in (i+1):K){
        iter <- iter + 1
        Sigma[i,j] <- rho[iter]*SD[i]*SD[j]
      }
    }
    Sigma <- Sigma + t(Sigma)
    diag(Sigma) <- SD^2
  }
  if(!all(eigen(Sigma)$values > 0))
    stop("matrix 'Sigma' must be positive definite")
  Sigma.cor <- cov2cor(Sigma)
  
  if(sig.level <= 0 | sig.level >= 1)
    stop("significance level must be in (0, 1)")
  
  ## calculations
  
  if(is.null(power)){
    std.effect <- delta/sqrt(diag(Sigma))
    z.alpha <- qnorm(1-sig.level)
    crit.vals <- z.alpha - sqrt(n/2)*std.effect
   power <- 1-pmvnorm(lower = -crit.vals, sigma = Sigma.cor) 
  }
  if(is.null(n)){
    std.effect <- delta/sqrt(diag(Sigma))
    z.alpha <- qnorm(1-sig.level)
    ssize.fct <- function(n, std.effect, z.alpha, Sigma.cor, power){
      crit.vals <- z.alpha - sqrt(n/2)*std.effect
      
      pmvnorm(lower = -crit.vals, sigma = Sigma.cor) - (1- power)
    }
    n <- uniroot(ssize.fct, c(2, 1e+05), tol = tol, extendInt = "yes",
                 std.effect = std.effect, z.alpha = z.alpha, Sigma.cor = Sigma.cor,
                 power = power)$root
  }
  NOTE <- "n is number in *each* group"
  METHOD <- "Power calculation for multiple primary endpoints for at least one endpoint"
  structure(list(n = n, delta = delta, SD = sqrt(diag(Sigma)),
                 rho = Sigma.cor[lower.tri(Sigma.cor)], Sigma = Sigma,
                 sig.level = sig.level,
                 power = power, note = NOTE, method = METHOD),
            class = "power.mpe.test")
}


mpe/R/power.known.var.R
power.known.var <- function(K, n = NULL, delta = NULL, Sigma, SD, rho, sig.level = 0.05,
                            power = NULL, tol = .Machine$double.eps^0.25){
  ## check of input
  if(missing(K))
    stop("specify the number of co-primary endpoints")
  if(!is.numeric(K))
    stop("'K' must be a natural number > 1")
  K <- as.integer(K)

  if(is.null(n) & is.null(power))
    stop("either 'n' or 'power' must be specified")
  if(!is.null(n) & !is.null(power))
    stop("either 'n' or 'power' must be NULL")
  if(!is.null(n)){
    if(length(n) > 1){
      warning("length of 'n' is greater than 1, only the first entry is used")
      n <- n[1]
    }
    n <- as.integer(n)
  }
  if(!is.null(power)){
    if(power <= 0 | power >= 1)
      stop("power must be in (0, 1)")
  }
  if(is.null(delta))
    stop("expected effect size 'delta' is missing")
  if(length(delta) != K)
    stop("length of 'delta' must be equal to 'K'")
  if(!all(delta > 0))
    stop("all effect sizes need to be positive")
  if(!missing(Sigma)){
    if(nrow(Sigma) != ncol(Sigma))
      stop("covariance matrix 'Sigma' must be quadratic")
    if(nrow(Sigma) != K)
      stop("covariance matrix must have dimension 'K' x 'K'")
    if(max(abs(Sigma - t(Sigma))) > 1e-10)
      stop("matrix 'Sigma' must be symmetric")
  }
  if(missing(Sigma)){
    if(missing(SD) || missing(rho))
      stop("if 'Sigma' is missing 'SD' and 'rho' must be given.")
    if(length(SD) != K)
      stop("length of 'SD' must be equal to 'K'")
    if(length(rho) != 0.5*K*(K-1))
      stop("length of 'rho' must be equal to '0.5*K*(K-1)'")
    Sigma <- matrix(0, nrow = K, ncol = K)
    iter <- 0
    for(i in 1:(K-1)){
      for(j in (i+1):K){
        iter <- iter + 1
        Sigma[i,j] <- rho[iter]*SD[i]*SD[j]
      }
    }
    Sigma <- Sigma + t(Sigma)
    diag(Sigma) <- SD^2
  }
  if(!all(eigen(Sigma)$values > 0))
    stop("matrix 'Sigma' must be positive definite")
  Sigma.cor <- cov2cor(Sigma)

  if(sig.level <= 0 | sig.level >= 1)
    stop("significance level must be in (0, 1)")

  ## calculations
  if(is.null(power)){
    std.effect <- delta/sqrt(diag(Sigma))
    z.alpha <- qnorm(1-sig.level)
    crit.vals <- z.alpha - sqrt(n/2)*std.effect
    power <- pmvnorm(upper = -crit.vals, sigma = Sigma.cor)
  }
  if(is.null(n)){
    std.effect <- delta/sqrt(diag(Sigma))
    z.alpha <- qnorm(1-sig.level)
    ssize.fct <- function(n, std.effect, z.alpha, Sigma.cor, power){
      crit.vals <- z.alpha - sqrt(n/2)*std.effect
      pmvnorm(upper = -crit.vals, sigma = Sigma.cor) - power
    }
    n <- uniroot(ssize.fct, c(2, 1e+05), tol = tol, extendInt = "yes",
                 std.effect = std.effect, z.alpha = z.alpha, Sigma.cor = Sigma.cor,
                 power = power)$root
  }
  NOTE <- "n is number in *each* group"
  METHOD <- "Power calculation for multiple co-primary endpoints (covariance known)"
  structure(list(n = n, delta = delta, SD = sqrt(diag(Sigma)),
                 rho = Sigma.cor[lower.tri(Sigma.cor)], Sigma = Sigma,
                 sig.level = sig.level,
                 power = power, note = NOTE, method = METHOD),
            class = "power.mpe.test")
}

print.power.mpe.test <- function (x, digits = getOption("digits"), ...) {
  cat("\n", x$method, "\n\n")
  note <- x$note
  Sigma <- x$Sigma
  x[c("method", "note", "Sigma")] <- NULL
  cat(paste(format(names(x), width = 15L, justify = "right"),
            format(x, digits = digits), sep = " = "), sep = "\n")
  cat("\nSigma =\n")
  print(Sigma)
  if (!is.null(note))
    cat("\n", "NOTE: ", note, "\n\n", sep = "")
  else cat("\n")
  invisible(x)
}
